File: src/lmjm/bordero_calculator.py
import dataclasses
from decimal import Decimal

from lmjm.model.batch_financial_result import BatchFinancialResult

Q = Decimal("0.0001")

GROSS_INTEGRATOR_PCT = Decimal("5.1")


@dataclasses.dataclass
class BorderoInput:
    housed_count: int
    mortality_count: int
    piglet_weight: Decimal
    pig_weight: Decimal
    total_feed: Decimal
    days_housed: int
    cap: Decimal
    map_value: Decimal
    price_per_kg: Decimal
    piglet_adjustment: Decimal
    carcass_adjustment: Decimal


def _q(value: Decimal) -> Decimal:
    """Quantize a Decimal to 4 decimal places."""
    return value.quantize(Q)


def calculate_bordero(inp: BorderoInput) -> BatchFinancialResult:
    """Pure function: computes all derived Borderô fields. Raises ValueError on invalid input."""
    if inp.housed_count == 0:
        raise ValueError("Housed count must be greater than zero")
    if inp.days_housed <= 0:
        raise ValueError("Days housed must be positive")

    # Farm data
    pig_count = inp.housed_count - inp.mortality_count

    # Carcass calculations
    carcass_yield_factor = _q((inp.pig_weight - Decimal("6.629")) / Decimal("1.289"))
    piglet_carcass_weight = _q(inp.piglet_weight * carcass_yield_factor)
    pig_carcass_weight = _q(inp.pig_weight * carcass_yield_factor)
    total_piglet_carcass = _q(piglet_carcass_weight * inp.housed_count)
    total_pig_carcass = _q(pig_carcass_weight * pig_count)
    total_carcass_produced = _q(total_pig_carcass - total_piglet_carcass)

    if total_carcass_produced <= 0:
        raise ValueError("Total carcass produced must be positive")

    # Feed conversion
    real_conversion = _q(inp.total_feed / total_carcass_produced)
    adjusted_conversion = _q(real_conversion + inp.piglet_adjustment + inp.carcass_adjustment)

    # Mortality
    real_mortality_pct = _q(Decimal(inp.mortality_count) / Decimal(inp.housed_count) * 100)
    adjusted_mortality_pct = real_mortality_pct

    # Integrator percentage adjustments (standard Borderô approach)
    mortality_adjustment_pct = _q(inp.map_value - adjusted_mortality_pct)
    conversion_adjustment_pct = _q(inp.cap - adjusted_conversion)
    integrator_pct = _q(GROSS_INTEGRATOR_PCT + mortality_adjustment_pct + conversion_adjustment_pct)

    # Financial result
    gross_income = _q(total_carcass_produced * inp.price_per_kg * integrator_pct / 100)
    net_income = gross_income

    # Performance
    daily_weight_gain = _q((inp.pig_weight - inp.piglet_weight) / Decimal(inp.days_housed))
    daily_carcass_gain = _q((pig_carcass_weight - piglet_carcass_weight) / Decimal(inp.days_housed))

    # Per-pig metrics
    gross_income_per_pig = _q(gross_income / Decimal(pig_count))
    net_income_per_pig = _q(net_income / Decimal(pig_count))

    return BatchFinancialResult(
        pk="",
        sk="",
        housed_count=inp.housed_count,
        mortality_count=inp.mortality_count,
        pig_count=pig_count,
        piglet_weight=inp.piglet_weight,
        pig_weight=inp.pig_weight,
        total_feed=inp.total_feed,
        days_housed=inp.days_housed,
        cap=inp.cap,
        map_value=inp.map_value,
        price_per_kg=inp.price_per_kg,
        gross_integrator_pct=GROSS_INTEGRATOR_PCT,
        carcass_yield_factor=carcass_yield_factor,
        piglet_carcass_weight=piglet_carcass_weight,
        pig_carcass_weight=pig_carcass_weight,
        total_piglet_carcass=total_piglet_carcass,
        total_pig_carcass=total_pig_carcass,
        total_carcass_produced=total_carcass_produced,
        real_conversion=real_conversion,
        piglet_adjustment=inp.piglet_adjustment,
        carcass_adjustment=inp.carcass_adjustment,
        adjusted_conversion=adjusted_conversion,
        daily_weight_gain=daily_weight_gain,
        daily_carcass_gain=daily_carcass_gain,
        real_mortality_pct=real_mortality_pct,
        adjusted_mortality_pct=adjusted_mortality_pct,
        mortality_adjustment_pct=mortality_adjustment_pct,
        conversion_adjustment_pct=conversion_adjustment_pct,
        integrator_pct=integrator_pct,
        gross_income=gross_income,
        net_income=net_income,
        gross_income_per_pig=gross_income_per_pig,
        net_income_per_pig=net_income_per_pig,
    )

File: frontend/src/lib/borderoCalculator.ts
// TypeScript port of bordero_calculator.py and cap_map_calculator.py
// Uses standard number type for live preview (backend uses Decimal for persistence precision)

const GROSS_INTEGRATOR_PCT = 5.1;

/** Round to 4 decimal places (mirrors Python Decimal.quantize("0.0001")) */
function q(value: number): number {
  return Math.round(value * 10000) / 10000;
}

// ---------------------------------------------------------------------------
// calculateBordero
// ---------------------------------------------------------------------------

export interface BorderoInput {
  housedCount: number;
  mortalityCount: number;
  pigletWeight: number;
  pigWeight: number;
  totalFeed: number;
  daysHoused: number;
  cap: number;
  mapValue: number;
  pricePerKg: number;
}

export interface BorderoResult {
  housedCount: number;
  mortalityCount: number;
  pigCount: number;
  pigletWeight: number;
  pigWeight: number;
  totalFeed: number;
  daysHoused: number;
  cap: number;
  mapValue: number;
  pricePerKg: number;
  grossIntegratorPct: number;
  carcassYieldFactor: number;
  pigletCarcassWeight: number;
  pigCarcassWeight: number;
  totalPigletCarcass: number;
  totalPigCarcass: number;
  totalCarcassProduced: number;
  realConversion: number;
  pigletAdjustment: number;
  carcassAdjustment: number;
  adjustedConversion: number;
  dailyWeightGain: number;
  dailyCarcassGain: number;
  realMortalityPct: number;
  adjustedMortality: number;
  mortalityAdjustmentPct: number;
  conversionAdjustmentPct: number;
  integratorPct: number;
  grossIncome: number;
  netIncome: number;
  grossIncomePerPig: number;
  netIncomePerPig: number;
}

/**
 * Pure function: computes all derived Borderô fields.
 * Returns null instead of throwing for invalid input (used in live preview).
 */
export function calculateBordero(inp: BorderoInput): BorderoResult | null {
  if (inp.housedCount === 0) return null;
  if (inp.daysHoused <= 0) return null;

  const pigCount = inp.housedCount - inp.mortalityCount;
  if (pigCount <= 0) return null;

  // Carcass calculations. Weights are carried at full precision; only the
  // final derived fields are rounded (matches the deterministic Borderô).
  const pigCarcassWeight = (inp.pigWeight - 6.629) / 1.289;
  const pigletCarcassWeight = inp.pigletWeight * 0.656807 - 1.315747;
  const carcassYieldFactor = inp.pigWeight !== 0 ? pigCarcassWeight / inp.pigWeight : 0;
  const totalPigletCarcass = pigletCarcassWeight * inp.housedCount;
  const totalPigCarcass = pigCarcassWeight * pigCount;
  const totalCarcassProduced = totalPigCarcass - totalPigletCarcass;

  if (totalCarcassProduced <= 0) return null;

  // Feed conversion
  const realConversion = inp.totalFeed / totalCarcassProduced;
  // Adjustments are computed (not user inputs).
  const pigletAdjustment = (22 - inp.pigletWeight) * 0.0125;
  const carcassAdjustment = (85 - pigCarcassWeight) * 0.0095;
  const adjustedConversion = realConversion + pigletAdjustment + carcassAdjustment;

  // Mortality
  const realMortalityPct = q((inp.mortalityCount / inp.housedCount) * 100);
  // Adjusted mortality rate: days-housed (age) correction added to the real rate.
  const adjustedMortality = q((130 - inp.daysHoused) * 0.0183 + realMortalityPct);

  // Integrator percentage adjustments
  const mortalityAdjustmentPct = q((inp.mapValue - adjustedMortality) / 5);
  const conversionAdjustmentPct = q((inp.cap - q(adjustedConversion)) * 10);
  const integratorPct = q(GROSS_INTEGRATOR_PCT + mortalityAdjustmentPct + conversionAdjustmentPct);

  // Financial result
  const grossIncome = q(q(totalCarcassProduced) * inp.pricePerKg * integratorPct / 100);
  const netIncome = q(grossIncome * 0.985);

  // Performance
  const dailyWeightGain = q((inp.pigWeight - inp.pigletWeight) / inp.daysHoused);
  const dailyCarcassGain = q((pigCarcassWeight - pigletCarcassWeight) / inp.daysHoused);

  // Per-pig metrics
  const grossIncomePerPig = q(grossIncome / pigCount);
  const netIncomePerPig = q(netIncome / pigCount);

  return {
    housedCount: inp.housedCount,
    mortalityCount: inp.mortalityCount,
    pigCount,
    pigletWeight: inp.pigletWeight,
    pigWeight: inp.pigWeight,
    totalFeed: inp.totalFeed,
    daysHoused: inp.daysHoused,
    cap: inp.cap,
    mapValue: inp.mapValue,
    pricePerKg: inp.pricePerKg,
    grossIntegratorPct: GROSS_INTEGRATOR_PCT,
    carcassYieldFactor: q(carcassYieldFactor),
    pigletCarcassWeight: q(pigletCarcassWeight),
    pigCarcassWeight: q(pigCarcassWeight),
    totalPigletCarcass: q(totalPigletCarcass),
    totalPigCarcass: q(totalPigCarcass),
    totalCarcassProduced: q(totalCarcassProduced),
    realConversion: q(realConversion),
    pigletAdjustment: q(pigletAdjustment),
    carcassAdjustment: q(carcassAdjustment),
    adjustedConversion: q(adjustedConversion),
    dailyWeightGain,
    dailyCarcassGain,
    realMortalityPct,
    adjustedMortality,
    mortalityAdjustmentPct,
    conversionAdjustmentPct,
    integratorPct,
    grossIncome,
    netIncome,
    grossIncomePerPig,
    netIncomePerPig,
  };
}

// ---------------------------------------------------------------------------
// computeCapMap
// ---------------------------------------------------------------------------

export interface CapMapResult {
  cap1: number;
  cap2: number;
  cap3: number;
  cap4: number;
  map1: number;
  map2: number;
}

/**
 * Computes CAP(1-4) and MAP(1-2) from integrator weekly data.
 */
export function computeCapMap(
  car: number,
  mar: number,
  avgSlaughterWeight: number,
  avgPigletWeight: number,
  averageAge: number,
): CapMapResult {
  const cap1 = q(car - (avgSlaughterWeight - 85) * 0.0095 - (avgPigletWeight - 22) * 0.0125);
  const cap2 = q(cap1 - 0.03);
  const cap3 = q(cap2 - 0.015);
  const cap4 = q(cap1 - 0.015);
  const map1 = q((130 - averageAge) * 0.0183 + mar);
  const map2 = q(map1 - 0.4);
  return { cap1, cap2, cap3, cap4, map1, map2 };
}

// ---------------------------------------------------------------------------
// selectCapMapVariant
// ---------------------------------------------------------------------------

export interface IntegratorWeeklyDataRecord {
  dateGenerated: string;
  validityStart: string;
  validityEnd: string;
  cap1: number;
  cap2: number;
  cap3: number;
  cap4: number;
  map1: number;
  map2: number;
}

/**
 * Selects the correct CAP and MAP variant based on batch characteristics.
 *
 * CAP selection:
 *   >2 origins + Creche → CAP(1)
 *   ≤2 origins + Creche → CAP(2)
 *   ≤2 origins + UPL    → CAP(3)
 *   >2 origins + UPL    → CAP(4)
 *
 * MAP selection:
 *   ≤2 origins → MAP(1)
 *   >2 origins → MAP(2)
 */
export function selectCapMapVariant(
  distinctOriginCount: number,
  predominantOriginType: "Creche" | "UPL",
  weeklyData: IntegratorWeeklyDataRecord,
): { cap: number; map: number } {
  let cap: number;
  if (predominantOriginType === "Creche") {
    cap = distinctOriginCount > 2 ? weeklyData.cap1 : weeklyData.cap2;
  } else {
    cap = distinctOriginCount > 2 ? weeklyData.cap4 : weeklyData.cap3;
  }

  const map = distinctOriginCount > 2 ? weeklyData.map2 : weeklyData.map1;

  return { cap, map };
}

// ---------------------------------------------------------------------------
// findNearestWeeklyData
// ---------------------------------------------------------------------------

/**
 * Finds the IntegratorWeeklyData record whose validity period contains the
 * target date. If none matches, returns the record with the nearest
 * dateGenerated to the target date.
 *
 * Returns undefined if the records array is empty.
 */
export function findNearestWeeklyData(
  records: IntegratorWeeklyDataRecord[],
  targetDate: string,
): IntegratorWeeklyDataRecord | undefined {
  if (records.length === 0) return undefined;

  // First: look for a record whose validity period contains the target date
  for (const record of records) {
    if (record.validityStart <= targetDate && targetDate <= record.validityEnd) {
      return record;
    }
  }

  // Fallback: find the record with the nearest dateGenerated
  let nearest = records[0];
  let minDiff = Math.abs(new Date(targetDate).getTime() - new Date(nearest.dateGenerated).getTime());

  for (let i = 1; i < records.length; i++) {
    const diff = Math.abs(new Date(targetDate).getTime() - new Date(records[i].dateGenerated).getTime());
    if (diff < minDiff) {
      minDiff = diff;
      nearest = records[i];
    }
  }

  return nearest;
}
